text/0000-async-await.md

```diff
@@ -0,0 +1,171 @@
+        create_v(t)
+    
+    fun create_v(t): V =>
+        let v: V = /* compute V */
+        v
+```
+
+> Future work might look at allowing behaviours to return `Promise` instances to make the above pattern and async / await even more friendly.
+
+There are some challenges regarding the capture of the lexical scope by the lambdas, but they can be solved by following the same rules as for lambdas: all captures must be explicit.
+
+Therefore, in the case of `await` calls, a capture would look similar to lambda calls:
+
+```pony
+class Foo
+    new create(env: Env, p: Promise[T]) =>
+        let t: T = await(env) p
+        // env can be used here as it was capture above
+        t.use(env)
+```
+
+De-sugared to Promises and lambdas, this would become:
+
+```pony
+class Foo
+    new create(env: Env, p: Promise[T]) =>
+        p.next[None]({(t: T)(env) =>
+            // env can be used here as it was capture above
+            t.use(env)
+        })
+```
+
+The implementation of this feature, for the above reasons, can and should be limited to the compiler's de-sugaring phase, without any other changes required.
+
+## Limitations
+
+`async` should be limited to function bodies only, not including constructors. Even though the single abstraction of asynchronous computation will continue to be Actor's behaviours, this limitation is necessary because:
+
+* the feature remains opt-in, so programmers have a choice on whehter they want to use it.
+* it is necessary to enforce that `async` functions return a `Promise` and no other types are allowed.
+* it should help avoid slowing down the compiler as most functions presumably will not be marked `async`.
+* behaviours cannot currently return values, so this feature doesn't make sense for them.
+
+# How We Teach This
+
+Because several popular languages have already introduced this pattern, the terminology used to describe it is pretty well established already. However, some of the terminology does not apply in Pony as well as it does in other languages.
+
+For example, async / await is commonly described as follows (taken from the [Wikipedia article](https://en.wikipedia.org/wiki/Async/await)):
+
+```
+... the async/await pattern is a syntactic feature of many programming languages that allows an asynchronous, non-blocking function to be structured in a way similar to an ordinary synchronous function.
+```
+
+what "blocking" and "non-blocking" mean may not be clear to everyone in the context of the Pony runtime. It may be best to avoid using such language, preferring only synchronous VS asynchronous execution.
+
+A call to `await` absolutely does not mean to block, it simply means to give up execution to another actor, a `Promise`, until it calls back (or errors, or even never) and executes the remaining of the body of the function (which is really, just a lambda inside the body of the function). This is why `async` functions are not allowed to return anything but a `Promise`.
```

Comment:
What happens to the actor that "gives up execution"? Is its stack saved? Does it continue processing messages? If yes, how do you reconcile the stack that was saved with the context of the might have changed (fields for example) due to processing of additional messages.

"give up execution to another actor" is rather vague. 

"it simply means to give up execution to another actor, a `Promise`, until it calls back (or errors, or even never) and executes the remaining of the body of the function (which is really, just a lambda inside the body of the function). This is why `async` functions are not allowed to return anything but a `Promise`."

isn't really how Pony works.

I don't think this is purely sugar. With a Promise, I give it a function to execute later when the promise is fulfilled and there is no local state that is saved for me to return to.

Are you suggesting that as part of the sugar, that all state on the stack is captured in a lambda that will be executed when the promise is fulfilled? But, with some sort of partial application?

Imagine for example:

```pony
let a: U64 = 1
let b: U64 = 2
let c: U64 = await ... something...
a + b +c
```

How does the above work?

Reply:
From one part of this it might be that I need to do this? It's not clear to me:

```pony 
let p: Promise[U64] = ... create some promise...
let a: U64 = 1
let b: U64 = 2
let c: U64 = await(a, b) p
a + b + c
```

Reply:
> What happens to the actor that "gives up execution"? Is its stack saved? Does it continue processing messages?

I don't understand these questions in light of my proposal.
Let me ask the same questions considering how Pony exists today:

* What happens to an actor that calls `next` on a Promise, "giving up execution" (because the behaviour ends there) until the lambda it passed into `next` is called? It its stack saved? Does it continue processing messages?

The answers with my proposal are the same as today.

```pony
let a: U64 = 1
let b: U64 = 2
let c: U64 = await ... something...
a + b +c
```

You need to capture `a` and `b` if you need to use them after await (just like today when you create a lambda):

```pony
let c: U64 = await(a, b) .... something
a + b + c
```

Then the whole thing becomes:

```pony
let a: U64 = 1
let b: U64 = 2
(some promise).next[T]({(c: U64)(a, b) =>
    a + b + c
})
```

Reply:
> What happens to an actor that calls next on a Promise

A promise is an actor. `next` is an asynchronous message send to it and function execution continues as per normal like it would after any other message send.